### PR TITLE
Add mainframe (aka SystemZ z/os) as a platform / architecture to MSF

### PR DESCRIPTION
**Pull request by bigendiansmalls**
This pull has the core files required to add the Mainframe platform (IBM's System Z) as an architecture / platform within MSF, so regular modules can then be written for it.

Nothing within this pull should have any affect whatsoever on existing modules.  It simply defines, where appropriate, the definitions for this platform/architecture.

This pull has the following items:
- Changes in `lib/msf/core`
  - `module/platform.rb` - added mainframe as a platform
  - `payload.rb` - required new type of payload - mainframe
  - `payload/mainframe.rb` - new file for basic mainframe payload
- Changes in `lib/rex`
  - `arch.rb` - added requirement for zarch as well as defined case for pack_addr
  - `arch/zarch.rb` - net new file for zarchitecture
  - `constants.rb` - added ARCH_ZARCH and included in ARCH_TYPES
  - `text.rb` - created two tables for ebcdic(cp1047)<->ascii conversion as well as the methods to use those tables
- Changes in `modules/exploits`
  - `multi/handler.rb` - inlcuded mainframe in the Platform array
- Changes in `spec/lib/rex`
  - `text_spec.rb` - added tests for to_ibm1047 and from_ibm1047 with the problem characters as tests
  - `arch/zarch_spec.rb` - added test for packing 64bit BE string (non-conversion)
# RSpec / Validation
- [x] $ rspec -f doc spec/lib/rex/text_spec.rb

```
Rex::Text
  Class methods
    .from_ebcdic
      should convert EBCDIC to ASCII (both US standards)
      should raise on non-convertable characters
    .to_hex
      should convert all chars 00 through ff
      should use the given prefix
    .rand_name
      should return a random name
    .gzip
      should return a properly formatted gzip file
    .to_octal
      should convert all chars 00 through ff
      should use the given prefix
    .to_ebcdic
      should convert ASCII to EBCDIC (both US standards)
      should raise on non-convertable characters
    .ungzip
      should return an uncompressed string
    .rand_name_male
      should return a random male name
    .randomize_space
      should return a string with at least one new space characater
      should not otherwise be mangled
    .to_hex_ascii
      should handle non-printables
      should not mess with printables
    .from_1047
      should convert mainframe EBCDIC (cp1047) to ASCII (ISO-8859-1)
    .rand_name_female
      should return a random female name
    .rand_mail_address
      should return a random mail address
    .rand_surname
      should return a random surname
    .to_utf8
      should convert a string to UTF-8, skipping badchars
    .to_ibm1047
      should convert ASCII to mainfram EBCDIC (cp1047)

Finished in 0.06846 seconds
22 examples, 0 failures
```
- [x]  $ rspec -f doc spec/lib/rex/arch/zarch_spec.rb 

```
Rex::Arch
  .pack_addr
    when arch is ARCH_ZARCH
      packs addr as 64-bit unsigned, big-endian

Finished in 0.03077 seconds
1 example, 0 failures

Randomized with seed 44844
```

And finally
- [x] in msfconsole  search platforms:mainframe should yield 2 results
  image

### CODE REVIEW

**Comment by bigendiansmalls** [comment]
FYI I noticed I'd inadvertently removed FIREFOX as a platform in constants.rb.  That's fixed in the most recent commit

File: modules/exploits/multi/handler.rb [hunk 1]
```diff
@@ -30,7 +30,7 @@ def initialize(info = {})
           'BadChars'    => '',
           'DisableNops' => true,
         },
-      'Platform'       => %w{ android bsd java js linux osx nodejs php python ruby solaris unix win },
+      'Platform'       => %w{ android bsd java js linux osx nodejs php python ruby solaris unix win mainframe },
```
Comment (bcook-r7) [inline modules/exploits/multi/handler.rb]: is 'mainframe' the right word here, or should it be more specific to allow for other mainframe architectures? Do such things exist that we would care about?

**Comment by bigendiansmalls** [comment]
Hey -- Well in keeping with the list, the oS is technically referred to now as z/os, but has changed names over the years several times (MVS, OS360, and more).   I think mainframe is a good catchall for these IBM produced systems. 

By my experience and also by standard industry vernacular mainframe == the IBM z series.  zseries is the latest marketing name and has changed several times over the years.  My thought was that mainframe was a good catchall for this system.  There might be one other system (made by Unisys) that might be lumped in here too - but if that platform ever makes its way onto the framework, I see no reason it could be part of that overall platform.  The architecture could be the delimiter if ever needed.

**Comment by bcook-r7** [comment]
Thanks for the explanation, and for keeping this compact and testable. Looking forward to subsequent PRs.

**Comment by bigendiansmalls** [comment]
You bet.   There's definitely more to come.

**Pull request merged**